File: room-demo-rxandroid-livedata/src/main/java/com/example/pavneet_singh/room_demo_rxandroid/notedb/NoteDatabase.java
package com.example.pavneet_singh.room_demo_rxandroid.notedb;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

import com.example.pavneet_singh.room_demo_rxandroid.notedb.dao.NoteDao;
import com.example.pavneet_singh.room_demo_rxandroid.notedb.model.Note;
import com.example.pavneet_singh.room_demo_rxandroid.util.Constants;
import com.example.pavneet_singh.room_demo_rxandroid.util.DateRoomConverter;

/**
 * Created by Pavneet_Singh on 12/31/17.
 */

@Database(entities = {Note.class}, version = 1)
@TypeConverters({DateRoomConverter.class})
public abstract class NoteDatabase extends RoomDatabase {

    public abstract NoteDao getNoteDao();


    private static NoteDatabase noteDB;

    // synchronized is use to avoid concurrent access in multithred environment
    public static /*synchronized*/ NoteDatabase getInstance(Context context) {
        if (null == noteDB) {
            synchronized (NoteDatabase.class) { // dual lock
                if (null == noteDB) {
                    noteDB = buildDatabaseInstance(context);
                }
            }
        }
        return noteDB;
    }

    private static NoteDatabase buildDatabaseInstance(Context context) {
        return Room.databaseBuilder(context,
                NoteDatabase.class,
                Constants.DB_NAME).build();
    }

    public void cleanUp() {
        noteDB = null;
    }
}

File: room-demo-kotlin-mvvm-dagger/src/main/java/com/example/pavneet_singh/room_demo_kotin_mvvm_dagger/notedb/NoteDataBase.kt
package com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.notedb

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.notedb.dao.NoteDao
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.notedb.model.Note
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.util.Constants
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.util.DateRoomConverter

/**
 * Created by Pavneet_Singh on 2020-01-25.
 */

@Database(
    entities = [Note::class],
    version = 1
)
@TypeConverters(
    DateRoomConverter::class
)
abstract class NoteDataBase : RoomDatabase() {
    abstract fun getNoteDao(): NoteDao

    companion object {
        private var noteDB: NoteDataBase? = null
        // synchronized is use to avoid concurrent access in multithred environment
        @JvmStatic
        fun  /*synchronized*/getInstance(context: Context): NoteDataBase {
            if (null == noteDB) {
                synchronized(NoteDataBase::class) {
                    // dual lock
                    if (null == noteDB) {
                        noteDB = buildDatabaseInstance(context)
                    }
                }
            }
            return noteDB!!
        }

        @JvmStatic
        private fun buildDatabaseInstance(context: Context): NoteDataBase {
            return Room.databaseBuilder(
                context,
                NoteDataBase::class.java,
                Constants.DB_NAME
            ).build()
        }
    }
}